Cython/Shadow.py

```diff
@@ -178,14 +181,13 @@ def typeof(arg):
 def address(arg):
     return pointer(type(arg))([arg])
 
-def declare(type=None, value=_Unspecified, **kwds):
-    if type not in (None, object) and hasattr(type, '__call__'):
-        if value is not _Unspecified:
-            return type(value)
-        else:
-            return type()
+def declare(t=None, value=_Unspecified, **kwds):
+    if value is not _Unspecified:
+        return cast(t, value)
+    elif (isinstance(t, type) and issubclass(t, (StructType, UnionType))) or isinstance(t, typedef):
+        return t()
```

Comment:
I find this change a bit questionable. I mean, it might be ok, it's just not obvious that this is the ultimate list of types that should be handled this way, and nothing else. I'd be happier if we could avoid introducing a regression here.

Reply:
@scoder  I totally understand your concern, but I don't see other solution which doesn't include the list of types. However, I don't feel like this solution is fragile as they are very special types. I'll try to explain my logic:
1. This branch is used for declaring a variable without any initializations i.e. `a = declare(cython.int)`
2. The result variable becomes uninitialized in case of simple types like `int` from C compiler perspective and it looks reasonable if it becomes `None` when interpreted
3. In the case of the classes, the value `None` matches the generated C code
4. The hardest case is structs, unions, and arrays. They are allocated on the stack in C and we cannot just return `None`. In such cases, we actually return the instance of the type witch exactly matches C code. That's why we need this check against the list of types 
5. The last case is the typedef. It is handled on the basis of the `base_type` (in the new commit which has been just added)

I've added more tests to verify everything. Please take a look and let me know what do you think. I'm totally flexible to change the implementation as long as all the new tests pass.